aten/src/ATen/cuda/detail/CUDAHooks.cpp

```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
   c10::cuda::device_synchronize();
 }
 
+std::tuple<size_t, size_t, ptrdiff_t, std::string, std::string, std::string, uint64_t, bool>
+CUDAHooks::StorageShareDevice(const c10::Storage& storage) const {
+    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+    auto shandle = c10::cuda::CUDACachingAllocator::shareIpcHandle(storage.mutable_data());
+    size_t ipc_memory_handle_size = shandle.handle.size();
+    size_t ipc_event_handle_size = sizeof(cudaIpcEventHandle_t);
+    std::string memory_handle(ipc_memory_handle_size, '\0');  // Using std::string instead of vector
+    std::string event_handle(ipc_event_handle_size, '\0');
+    std::string ref_counter(sizeof(std::string), '\0');
+    uint64_t ref_counter_offset;
+    bool event_sync_required;
+    ptrdiff_t offset_bytes = shandle.offset;
+    std::memcpy(memory_handle.data(), shandle.handle.c_str(), shandle.handle.size());
+
+    // Put Storage Data behind new ref counting context
+    // See Note [CUDA IPC Refcounting implementation explained]
+    at::DataPtr sent_data_ptr = torch::GetNewRefCountedSentData(
+        storage.mutable_data(), storage.device());
+    auto old_data_ptr = storage.set_data_ptr(std::move(sent_data_ptr));
+    auto sent_data =
+        static_cast<torch::CudaIPCSentData*>(storage.data_ptr().get_context());
+    sent_data->set_original_ptr(std::move(old_data_ptr));
+    std::memcpy(ref_counter.data(), (sent_data->handle()).c_str(), sizeof(sent_data->handle()));
+    ref_counter_offset = sent_data->offset();
+
+    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+    cudaIpcEventHandle_t ipc_event_handle;
+    if (sent_data->event_sync_required_) {
+        C10_CUDA_CHECK(
+            cudaIpcGetEventHandle(&ipc_event_handle, sent_data->event_));
+    }
+    std::memcpy(event_handle.data(), (char*)&ipc_event_handle, sizeof(cudaIpcEventHandle_t));
+    event_sync_required = sent_data->event_sync_required_;
+
+    // Return the results as a tuple
+    return std::make_tuple(ipc_memory_handle_size, ipc_event_handle_size,
+                           offset_bytes, std::move(memory_handle),
+                           std::move(event_handle), std::move(ref_counter),
+                           ref_counter_offset, event_sync_required);
+}
+
+
+void CUDAHooks::StorageNewSharedDevice(const c10::DeviceIndex& device,
+                                       bool& event_sync_required,
+                                       std::string& s_ipc_event_handle,
+                                       std::string& s_handle,
+                                       std::string& ref_counter_handle,
+                                       ptrdiff_t& ref_counter_offset,
+                                       ptrdiff_t& storage_offset_bytes,
+                                       c10::DataPtr& data_ptr) const {
```

Comment:
You only have one return, return this by value lol

Reply:
@ezyang Modified